File: 3904_Smallest_Stable_Index_II/sol.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 * Time:    O(N)
 * Space:   O(N)
 * Strategy: Prefix & suffix
 */
var func = function(nums, k) {
    // Approach 2
    let curr_max = nums[0];
    let suffixmin = new Array(nums.length);
    suffixmin[nums.length-1] = nums.at(-1);
    for (let i = nums.length-2; i >= 0; i--) {
        suffixmin[i] = Math.min(suffixmin[i+1], nums[i])
    }
    for (let i = 0; i < nums.length; i++) {
        curr_max = Math.max(curr_max, nums[i]);
        if (k >= curr_max- suffixmin[i]){
            return i;
        }
    }return -1;

    // Approach 1 
    // let curr_max = nums[0];
    // let curr_min = Math.min(...nums);
    // for (let i = 0; i < nums.length; i++) {
    //     curr_max = Math.max(curr_max, nums[i]);
    //     if (k >= curr_max- curr_min){
    //         return i;
    //     }
    //     if (curr_min == nums.at(i)){
    //         curr_min = Math.min(...nums.slice(i+1));
    //     }
    // }
    // return -1;
};
/**
 * Local testing
 */
function test() {
    const testCases = [
        { args: [[5,0,1,4], 3], expected: 3 },
        { args: [[3,2,1], 1], expected: -1 },
        { args: [[0], 0], expected: 0 }
    ];
    testCases.forEach(({ args, expected }, index) => {
        const result = func(...args);
        const status = (result === expected) ? "✅ PASS" : "❌ FAIL";
        console.log(
            `${status} | Test #${index + 1}: ` +
            `Input: ${JSON.stringify(args)} | ` +
            `Expected: ${expected} | ` +
            `Result: ${result}`
        );
    });
}
test();

File: 3904_Smallest_Stable_Index_II/sol.java
import java.util.*;

class Solution {
    /**
     * Time:    O(N)
     * Space:   O(N)
     * Strategy: Prefix & Suffix
     */
    public int solve(int[] nums, int k) {
        int curr_max = nums[0];
        int[] suffix_min = new int[nums.length];
        suffix_min[nums.length -1] = nums[nums.length -1];
        for(int i = nums.length-2; i >= 0; i--){
            suffix_min[i] = Math.min(suffix_min[i+1], nums[i]);
        }
        for(int i=0; i< nums.length; i++){
            curr_max = Math.max(curr_max, nums[i]);
            if (k>= curr_max - suffix_min[i]) {
                return i;
            }
        }
        return -1;
    }
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[][] inputs = {
            {5,0,1,4},
            {3,2,1},
            {0}
        };
        int[] targets = {3, 1, 0};
        int[] expected = {3, -1, 0};
        System.out.println("Running Tests...");
        for (int i = 0; i < inputs.length; i++) {
            int result = sol.solve(inputs[i], targets[i]);
            String status = (result == expected[i]) ? "✅ PASS" : "❌ FAIL";
            System.out.printf("%s | Test #%d: Input: %s, Target: %d | Expected: %d | Result: %d%n",
                status, 
                i + 1, 
                Arrays.toString(inputs[i]), 
                targets[i], 
                expected[i], 
                result
            );
        }
    }
}